File: backend/src/main/java/com/d4k/ecommerce/modules/review/service/ReviewService.java
package com.d4k.ecommerce.modules.review.service;

import com.d4k.ecommerce.modules.review.dto.request.ReviewRequest;
import com.d4k.ecommerce.modules.review.dto.response.ReviewResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * Review Service Interface
 * Định nghĩa business logic cho reviews
 */
public interface ReviewService {
    
    /**
     * Tạo review mới
     * @param userId user ID
     * @param request thông tin review
     * @return review đã tạo
     */
    ReviewResponse createReview(Long userId, ReviewRequest request);
    
    /**
     * Lấy danh sách reviews của một product
     * @param productId product ID
     * @param pageable thông tin phân trang
     * @return danh sách reviews
     */
    Page<ReviewResponse> getProductReviews(Long productId, Pageable pageable);
    
    /**
     * Lấy reviews của user
     * @param userId user ID
     * @param pageable thông tin phân trang
     * @return danh sách reviews
     */
    Page<ReviewResponse> getUserReviews(Long userId, Pageable pageable);
    
    /**
     * Xóa review
     * @param reviewId review ID
     * @param userId user ID (để check ownership)
     */
    void deleteReview(Long reviewId, Long userId);
    
    /**
     * Admin xóa review
     * @param reviewId review ID
     */
    void deleteReviewByAdmin(Long reviewId);
    
    /**
     * Lấy rating trung bình của product
     * @param productId product ID
     * @return average rating
     */
    Double getAverageRating(Long productId);
    
    /**
     * Lấy số lượng reviews của product
     * @param productId product ID
     * @return review count
     */
    Long getReviewCount(Long productId);
}

File: frontend/src/services/review-service.js
import apiClient from './api-client';

/**
 * Review Service
 * Các API liên quan đến reviews
 */
const reviewService = {
  /**
   * Lấy reviews của sản phẩm
   * @param {Number} productId - Product ID
   * @param {Object} params - Query parameters (page, size)
   * @returns {Promise}
   */
  getProductReviews: (productId, params = {}) => {
    return apiClient.get(`/reviews/product/${productId}`, { params });
  },

  /**
   * Tạo review mới
   * @param {Object} reviewData - { productId, rating, comment }
   * @returns {Promise}
   */
  createReview: (reviewData) => {
    return apiClient.post('/reviews', reviewData);
  },

  /**
   * Xóa review
   * @param {Number} reviewId - Review ID
   * @returns {Promise}
   */
  deleteReview: (reviewId) => {
    return apiClient.delete(`/reviews/${reviewId}`);
  },

  /**
   * Lấy review của user cho sản phẩm
   * @param {Number} productId - Product ID
   * @returns {Promise}
   */
  getUserReview: (productId) => {
    return apiClient.get(`/reviews/my-review/${productId}`);
  },
};

export default reviewService;

